### PR TITLE
Improve behavior of AbstractLatentWorker by implementing a state machine

### PR DESCRIPTION
**Pull request by p12tic**
Previously the latent worker tracked its state implicitly by looking into the following variables: self.conn, self._substantiation_notifier, self.insubstantiating. As a result, the implementation was seemingly simple, but it didn't handle the edge cases well. For example, in a lot of cases `stop_instance()` was called incorrect number of times and integration tests didn't even check that. The latent worker implementations were less robust as a result (e.g. causing issues like #4412).

This PR switches `AbstractLatentWorker` to a state machine and additionally enables proper tracking of latent worker state in its integration tests. This has led to more robust behavior which is covered by 10 new latent worker integration tests. We now document what the LatentWorker implementations may expect in terms of guarantees of when `start_instance()` and `stop_instance()` is called.

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/test/fake/latent.py [hunk 1]
```diff
@@ -42,15 +42,18 @@ class LatentController(SeverWorkerConnectionMixin):
     stop_instance() is executed.
     """
 
+    STATE_NOT_STARTED = 0
```
Comment (tardyp) [inline master/buildbot/test/fake/latent.py]: would it make sense to use automat to implement this?

https://pypi.org/project/Automat/

Comment (p12tic) [reply master/buildbot/test/fake/latent.py]: I think that using automat would make code less readable. It could probably work for `LatentController`, but in `AbstractLatentWorker` the '.state' member does not represent entire state and we make decisions about the next state taking into account more data than that. I'm probably biased, but in this case procedural paradigm allows easier building of the mental model than a declarative one.

By the way, I saw somewhere within Buildbot state constants being represented in a separate class as `NamedConstant` members. Should I do that here too?

Comment (seankelly) [reply master/buildbot/test/fake/latent.py]: Since it's 3.5+, what about using enum?

Comment (p12tic) [reply master/buildbot/test/fake/latent.py]: OK, will use enum.

Comment (p12tic) [reply master/buildbot/test/fake/latent.py]: Should be fixed now.

**Comment by p12tic** [comment]
@tardyp: Are there any issues or risks with this PR that still need to be addressed before merging?
